Add support for pinning snapshots in gradle resolver

acecilia opened the pull request:
Hi 👋 

This PR adds support for snapshot dependencies in the gradle resolver. It uses the repository timestamp of the snapshot as a version. I added this information in a new field with a generic name `VersionRevision`.

Related: https://github.com/bazel-contrib/rules_jvm_external/pull/974

acecilia commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/remote/Downloader.java:
```diff
@@ -84,11 +84,7 @@ public DownloadResult download(Coordinates coords) {
-    String pomName = String.format("%s-%s.pom", coords.getArtifactId(), coords.getVersion());
-    String pom = Paths.get(originalTarget).getParent().resolve(pomName).toString();
-
+    String pom = coords.setExtension("pom").toRepoPath();
```
Not sure why the special handling of POM files above 🤔

acecilia commented on tests/custom_maven_install/regression_testing_gradle_install.json:
```diff
@@ -243,6 +243,37 @@
+        "jar": "79abe7953803b20d1deb5f36283d4d640c7aacbdb1ddd7601aae369b1633e903"
+      },
+      "version": "999.0.0-HEAD-jre-SNAPSHOT",
+      "version_revision": "999.0.0-HEAD-jre-20250623.150948-114"
```
Here is now the snapshot version pinning looks like, including the timestamp

acecilia commented:
Thanks for the feedback. Let me ask for some more context to understand a bit better your thoughts 😄 

> Snapshots aren't always stored with a version. Sometimes, the -SNAPSHOT.jar is just silently replaced

By "version", do you mean "timestamp"? If yes, is there maybe a public snapshot you could share that I could use during development, and also to add tests?

> We should have a consistent story for handling snapshots throughout the ruleset

What do you mean here, that snapshots should be handled by all resolvers? If yes, I agree. I was planning to look into the other resolvers after merging this PR. Is that okey, or would you prefer do do it differently?

> When we create the lockfile, omit the sha256 of any snapshot jar that doesn't have a fixed version (likely still contains SNAPSHOT.jar in the file name). That means that if the version is updated in place, builds won't break unexpectedly.

Let me expand a bit on my comment [here](https://github.com/bazel-contrib/rules_jvm_external/issues/901#issuecomment-3061649472). How would this work in bazel?
* Bazel uses `http_file` rule in [here](https://github.com/bazel-contrib/rules_jvm_external/blob/d61367cc18c6178838655446dbcdcb76c1bf4b7d/private/extensions/download_pinned_deps.bzl#L38) to download the artifacts. That rule has a sha256 property that is optional, but the [docs](https://bazel.build/rules/lib/repo/http#http_file-sha256) mention that _It is a security risk to omit the SHA-256 as remote files can change. At best omitting this field will make your build non-hermetic_ - we would be accepting these risks when adding support for such kind of snapshots
* For `http_file` rules without sha256, bazel would keep the downloaded artifacts in the cache, and will never try to update them. How do you envision that someone would trigger a snapshot update in this case? My understanding is that gradle caches snapshots but still fetches them once a day, but that is not the case for bazel
* My current understanding is that time-stamping snapshots is the default behaviour for maven repositories. I acknowledge that when searching in google "maven snapshots without timestamp" there are some ways to do it mentioned (example [here](https://maven.apache.org/plugins-archives/maven-deploy-plugin-2.6/examples/disabling-timestamps-suffix.html) and [here](https://stackoverflow.com/questions/1243574/how-to-stop-maven-artifactory-from-keeping-snapshots-with-timestamps)), but I also observed that those mentions are usually +10 years old. Furthermore, when trying to find more recent resources about how to do this, I found [this](https://stackoverflow.com/a/70653852) from 3 years ago that mentions that _Support for uniqueVersion was deprecated in Maven 2.x, and removed entirely in Maven 3_, which is confirmed by maven documentation in [here](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=2852668#Maven3.xCompatibilityNotes-Non-uniqueSnapshotDeployments). 

This way of integrating snapshots goes against bazel fundamentals + it is deprecated. This is what makes me question wether this ruleset wants/needs to support it. Wdyt?

acecilia commented:
@shs96c let me know 🙏

acecilia commented:
Got it thanks for that detailed answer 🙏  Added support for non-versioned snapshots in gradle resolver

acecilia commented on tests/bazel_run_tests.sh:
```diff
@@ -367,8 +379,10 @@ TESTS=(
   "test_when_both_pom_and_jar_artifact_are_available_jar_artifact_is_present"
   "test_when_both_pom_and_jar_artifact_are_dependencies_jar_artifact_is_present"
-  # "test_gradle_metadata_is_resolved_correctly_for_aar_artifact"
+  "test_gradle_metadata_is_resolved_correctly_for_aar_artifact"
```
For context about this change see [here](https://github.com/bazel-contrib/rules_jvm_external/pull/1395/files#r2264116068)

acecilia commented on private/rules/v2_lock_file.bzl:
```diff
@@ -138,6 +139,7 @@ def _get_artifacts(lock_file_contents):
             "group": parts[0],
             "artifact": parts[1],
             "version": data["version"],
+            "version_revision": data.get("version_revision"),
```
I added `version_revision` separately because I did not consider version and snapshot timestamp to be the same:
* `version`: this is the version obtained by calling `component.getModuleVersion().getVersion()` in [here](https://github.com/bazel-contrib/rules_jvm_external/pull/1412/files)
* `version_revision`: the timestamp (or revision) for a specific version

You can see a sample of how both look like in [here](https://github.com/bazel-contrib/rules_jvm_external/pull/1412/files#r2199375341).

Would you prefer I try to get rid of `version_revision`, merging both concepts into one?

acecilia replied:
Hi, found some time to look into this. I tried following the approach you proposed: getting rid of `version_revision` and using existing `version` instead. 

I was unable to make it work because the `version` field and the `version_revision` are needed separately.

This is my understanding of where the difference is important:

* After the lockfile is created, the [Downloader.java](https://github.com/bazel-contrib/rules_jvm_external/blob/1c5cfbf96de595a3e23cf440fb40380cc28c1aea/private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/remote/Downloader.java#L177) downloads the `jar` files
* The download url for the `jar` files is calculated from the information in the lockfile. The url is passed to the downloader in [here](https://github.com/bazel-contrib/rules_jvm_external/blob/1c5cfbf96de595a3e23cf440fb40380cc28c1aea/private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/remote/Downloader.java#L194), and is calculated by the [toRepoPath](https://github.com/bazel-contrib/rules_jvm_external/blob/1c5cfbf96de595a3e23cf440fb40380cc28c1aea/private/tools/java/com/github/bazelbuild/rules_jvm_external/Coordinates.java#L126) method
* The URL format is: 
    * `<base_url>/<getGroupId>/<artifactId>/<version>/<artifactId>/<version>` - for non-snapshots
    * `<base_url>/<getGroupId>/<artifactId>/<version>/<artifactId>/<snapshotId>` for versioned snapshots (note that `snapshotId` in the context of this PR is `versionRevision`)

So for the guava snapshots, the download URL is: `https://oss.sonatype.org/content/repositories/snapshots/com/google/guava/guava/999.0.0-HEAD-jre-SNAPSHOT/guava-999.0.0-HEAD-jre-20250623.150948-114.jar` (you can paste it in the browser and you will see it successfully downloads the `jar` file)

If I use a single `version` to store both the snapshot version and the timestamp version, then the URL is incorrect: `https://oss.sonatype.org/content/repositories/snapshots/com/google/guava/guava/999.0.0-HEAD-jre-20250623.150948-114/guava-999.0.0-HEAD-jre-20250623.150948-114.jar`

H5-O5 replied:
> Why is this needed? Surely the version is one of:
> 
> 1. A regular version number for non-snapshot deps
> 2. `x.x.x-SNAPSHOT` for snapshot deps that aren't versioned
> 3. A snapshot timestamp for snapshot deps that are versioned
> 
> Looking at the maven code, it doesn't look like their abstractions use `version_revision` either.

I'd like to join the discussion here. I'm in favor of adding a new field.

The url is `<baseUrl>/<groupId>/<artifactId>/<version>/<artifactId>-<version>`, e.g. `https://oss.sonatype.org/content/repositories/snapshots/com/google/guava/guava/999.0.0-HEAD-jre-SNAPSHOT/guava-999.0.0-HEAD-jre-20250623.150948-114.jar`

Currently we do:

```java
    path.append(getGroupId().replace('.', '/'))
        .append("/")
        .append(getArtifactId())
        .append("/")
        .append(getVersion())
        .append("/")
        .append(getArtifactId())
        .append("-")
        .append(getVersion());
```

But this code only works if the version and revision is the same.

If we keep using single field, then it must be guava-999.0.0-HEAD-jre-20250623.150948-114.jar. We'll have to add codes like this:

```
if version looks like a snapshot version, then
  real_version = remove the timestamp and append "-SNAPSHOT"
else
  real_version = just the version
```

And as the codebase evolves, we may need to add such a hack everywhere. and the codebase is harder to maintain. It's just easier to use 2 fields to remember 2 parts of the URL, especially the first part "xxx-SNAPSHOT" and the second part "xxx-TIMESTAMP" are different.

acecilia replied:
> I'd like to join the discussion here.

Appreciate it 🙌 

> At first I thought this is unnecessary and maintaining a new field is clearer

Yes, this is my thought also: having to decode `Coordinates.version` in all the places where it is used seems to me an unattainable and error-prone task.

> Upon further look I saw that in many places we have the equivalent of new Coordinates(old_coord.toString()). If we use 2 fields then there is no way to encode the same version-revision information in <groupId>:<artifactId>[:<extension>[:<classifier>][:<version>], but replacing version with revision and to part it on-the-fly can make the serialization and deserialization work as before.

Im sorry, I tried, but I do not understand what you mean here, I think I am missing some context. Could you maybe share an example that helps me grasp a bit better the comment 🙏

acecilia replied:
Updated as suggested: removed `version_revision` and instead encode the timestamp in the `version` field

acecilia commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/plugin/GradleDependencyModelBuilder.java:
```diff
@@ -268,7 +268,17 @@ private GradleResolvedDependency walkResolvedComponent(
+
+      // For snapshot dependencies, extract the timestamped version from the component ID
+      // Note: this is unsafe, is there a better way of obtaining the timestamp for an snapshot? I could not find any
+      if (component.getId().getClass().getName().contains("MavenUniqueSnapshotComponentIdentifier")) {
```
I tried to update it as suggested: using regex to identify snapshots. I could not use the exact regex tho, but very similar

acecilia commented on private/rules/coursier.bzl:
```diff
@@ -597,7 +597,9 @@ def _pinned_coursier_fetch_impl(repository_ctx):
-                "        sha256 = \"%s\"," % artifact["sha256"],
+                # sha256 is optional: non-versioned snapshots may not have it
+                # See: https://github.com/bazel-contrib/rules_jvm_external/pull/1412
+                "        sha256 = \"%s\"," % artifact.get("sha256", None),
```
Updated as suggested, thanks!

acecilia commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/models/GradleResolvedDependency.java:
```diff
@@ -34,6 +34,10 @@ public interface GradleResolvedDependency {
 
   void setVersion(String version);
 
+  String getVersionRevision();
```
See [here](https://github.com/bazel-contrib/rules_jvm_external/pull/1412#discussion_r2312126617) please

acecilia commented:
I am keen to complete this PR with your guidance, have a look when you have a chance please 🙏

smocherla-brex commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/gradle/GradleResolver.java:
```diff
@@ -275,7 +276,8 @@ private void addDependency(
                   extension,
                   childCoordinates.getClassifier(),
-                  childCoordinates.getVersion());
+                  childCoordinates.getVersion(),
```
What I was hoping here was substitute the versionRevision into the `version` field of `Coordinates` instead of the same version (if it was a snapshot one). Then in the downloader, it should try to download the actual revision of the snapshot? That way, you probably still need versionRevision in the gradle plugin not in `Coordinates` itself. I can try testing that query quickly as well. @acecilia

H5-O5 replied:
@acecilia I have an implementation of supporting snapshot for maven-resolver. I used the single field as @smocherla-brex suggested. The key is that we store the snapshot version (2025xxxx.xxxxxx-y) in Coordinates, and when downloading, calculate the real url. Have a look at https://github.com/bazel-contrib/rules_jvm_external/compare/master...H5-O5:rules_jvm_external:maven-snapshot and private/tools/java/com/github/bazelbuild/rules_jvm_external/Coordinates.java and it may help you get your MR aligned with the reviewer's request.

smocherla-brex replied:
I took a shot to try to use just the single version field in `Coordinates`, it requires additional handling within Coordinates but it does seem to work https://github.com/smocherla-brex/rules_jvm_external/commit/6b6a293c9340b342226c3c03a5895ff6bd1ca3cb#diff-0884e6cc400f368585972b32d3105196826dc1bdfb803843798f30b24ecc87c1R370-R371 (this is on top of this branch/PR) and making sure we handle it in toRepoPath(). I have not tested everything though. I'm not necessarily opposed to a new field but it could increase the lockfile size potentially and I'll defer to Simon on the decision.

H5-O5 replied:
@shs96c what's the status of this MR?

acecilia replied:
Thanks @shs96c for https://github.com/acecilia/rules_jvm_external/pull/1!

Now I hope I got it. Updated as suggested: removed version_revision and instead encode the timestamp in the version field

smocherla-brex commented on MODULE.bazel:
```diff
@@ -764,12 +764,17 @@ dev_maven.install(
+        # Versioned snapshot pinning support: https://github.com/bazel-contrib/rules_jvm_external/pull/1412
+        "com.google.guava:guava:999.0.0-HEAD-jre-SNAPSHOT",
+        # Non-versioned snapshot pinning support: https://github.com/bazel-contrib/rules_jvm_external/pull/1412
+        "org.seleniumhq.selenium:selenium-java:4.34.0-SNAPSHOT",
```
You'll need to port this change here as well https://github.com/bazel-contrib/rules_jvm_external/blob/master/WORKSPACE#L390 in WORKSPACE to resolve the CI failures related to Bazel 6.

acecilia replied:
Thanks added

acecilia commented on tests/custom_maven_install/regression_testing_gradle_install.json:
```diff
@@ -422,6 +525,114 @@
+      "shasums": {
+        "jar": null
+      },
+      "version": "4.34.0-SNAPSHOT"
```
Mmm I think this is because is selenium: the snapshot is non-versioned.

This PR contains 2 integrations:
* `guava`: versioned snapshot - with timestamps and `versionRevision`
* `selenium`: non-versioned snapshot - without timestamp nor `versionRevision`

Does it make sense now? Or am I missing something?

smocherla-brex replied:
Got it. I was under the assumption that snapshot versions always mapped to a timestamp version. Makes sense.

shs96c commented:
Apologies for the long delay getting back to you. I've been traveling and sick. I want to review the `version_revision` as it's making me uneasy, and I'd like to be satisfied it's required.

acecilia committed:
Fix KMP artifacts getting null sha256 when only POM is known

When Gradle resolves a Kotlin Multiplatform (KMP) module like
com.squareup.okio:okio, the JVM variant has no jar (the actual jar
lives in a separate okio-jvm module). The only file Gradle provides
for the root okio coordinate is a .pom file.

Commit 723d203 added an isPomPathForNonPomCoordinates check in
Downloader.performDownload that short-circuits when the known path is
a POM for a non-POM coordinate, returning DownloadResult with null
sha256 without trying to download the actual jar from the repos.

This caused okio to get null sha256 in the regenerated lockfile,
making it a java_library instead of jvm_import.

723d203 passed CI because the test at that time only checked that the
okio target exists (via plain 'bazel query'), not its rule kind. With
null sha256, okio is java_library but still exists, so the test passed.
This branch tightened the test to use 'bazel query --output=label_kind'
and expect 'jvm_import rule', which exposed the null sha256 bug.

The fix: instead of returning immediately, skip the POM known-path
and fall through to the download loop so the real jar can be fetched
from the remote repos. If the jar download succeeds, its sha256 is
computed normally. If it fails (e.g. jar genuinely missing), return
the POM-only result (null sha256) as before.

Also adds a new test verifying that when the jar IS available on the
repo, the downloader fetches it and returns a real sha256.

acecilia commented on private/tools/java/com/github/bazelbuild/rules_jvm_external/resolver/remote/Downloader.java:
```diff
@@ -174,6 +181,11 @@ private DownloadResult performDownload(Coordinates coordsToUse, String path) {
+      // POM-only result rather than null so the caller can still use the POM.
+      if (hadPomOnlyKnownPath) {
+        return new DownloadResult(coordsToUse, Set.of(), null, null);
+      }
```
Check the commit message in https://github.com/bazel-contrib/rules_jvm_external/pull/1412/changes/c1d50b096ceb9be8b3e43ce073691e6c114c645d for an explanation of this change (is fixing an existing bug introduced in the last master commit)

acecilia commented on tests/custom_maven_install/regression_testing_gradle_install.json:
```diff
@@ -440,11 +446,35 @@
+      "shasums": {
+        "jar": "9f71d9a756937d784cd175af7e3a1992f088321435b880af02488003da4c43b2"
+      },
+      "version": "999.0.0-HEAD-jre-20260701.164618-384"
```
Here is how a timestamped snapshot dependency shows: with the timestamp as part of the version

acecilia committed:
Add support for pinning snapshots in gradle resolver

acecilia commented on tests/custom_maven_install/regression_testing_gradle_install.json:
```diff
@@ -440,11 +446,35 @@
+      "shasums": {
+        "jar": null
+      },
+      "version": "999.0.0-HEAD-jre-20260828.195817-455"
```
Here is how a timestamped snapshot dependency shows: with the timestamp as part of the version

acecilia commented on tests/custom_maven_install/non_timestamped_snapshot_testing_gradle_install.json:
```diff
@@ -0,0 +1,32 @@
+  "artifacts": {
+    "com.example:snapshot": {
+      "shasums": {
+        "jar": null
```
Here is how a non-timestamped snapshot dependency shows: with the shasum set to null

acecilia commented on tests/custom_maven_install/timestamped_snapshot_testing_gradle_install.json:
```diff
@@ -0,0 +1,32 @@
+      "shasums": {
+        "jar": "07c82b630621b46fe45aff0e43de98feea968c9cb793f8229b91f12c5ef1e976"
+      },
+      "version": "1.0-20260829.120000-1"
```
Here is how a timestamped snapshot dependency shows: with the timestamp as part of the version
